site/js/runner.js

```diff
@@ -62,68 +49,62 @@ var Runner = {
             $( '<li class="state">&nbsp;</li>' )
         );
         $( '.runner' ).show();
-
-        this.displayRunStatus();
-        this.enabled = true;
-        this.editor.renderer.requestRendering();
     },
     run: function() {
         this.editor.inputSubmit();
-        this.inputString = prompt( 'Enter input string: ', 'abba' );
-
-        if ( typeof this.inputString != 'string' ) {
-            return false;
-        }
-        for ( var i = 0; i < this.inputString.length; ++i ) {
-            if ( !( this.inputString[ i ] in nfaview.nfa.alphabet ) ) {
+        var s = prompt( 'Enter input string: ', this.inputString ) + '';
+        for ( var i = 0; i < s.length; ++i ) {
+            if ( !( s[ i ] in nfaview.nfa.alphabet ) ) {
+                alert( 'Input string must contain only symbols from the alphabet' );
                 return false;
             }
         }
-
-        this.begin();
+        this.inputString = s;
+        this.editor.run( this.inputString );
+        this.open();
+        this.enabled = true;
+        this.runStep = 0;
+        this.step( 0 )
         return false;
     },
-    next: function() {
-        ++this.runStep;
-        if ( this.runStep > this.inputString.length ) {
-            this.runStep = this.inputString.length;
-            return false;
-        }
+    step: function ( step ) {
+        if ( step < 0 ) { step = 0; }
+        if ( step <= this.inputString.length ) {
+            var oldStep = this.runStep;
+            this.runStep = this.editor.gotoStep( this.inputString, step );
 
-        if ( this.editor.runStep() ) {
-            this.displayRunStatus();
-            this.displayAcceptanceResult();
-        }
-        else {
-            this.displayRejectedStatus();
+            if ( this.runStep == this.inputString.length ) {
+                if ( this.isAccepted() ) {
+                    this.displayAcceptedStatus();
+                } else {
+                    this.displayRejectedStatus();
+                }
+            } else {
+                if ( oldStep != this.runStep || oldStep == step ) {
+                    this.displayRunStatus();
+                } else {
+                    this.displayRejectedStatus();
+                }
+            }
+            this.displayStep();
+            this.editor.renderer.requestRendering();
         }
-        this.editor.renderer.requestRendering();
+        return false;
+    },
+    next: function() {
+        this.step( this.runStep + 1 );
         return false;
     },
     rewind: function() {
-        this.begin();
+        this.step( 0 );
         return false;
     },
     prev: function() {
-        var step = Math.max( 0, this.runStep - 1 );
-
-        this.begin();
-
-        this.runStep = step;
-
-        this.editor.gotoStep( this.inputString, this.runStep );
-        this.displayRunStatus();
-        this.editor.renderer.requestRendering();
-
+        this.step( this.runStep - 1 );
```

Comment:
Not sure why we can get rid of the boundary condition check here. Why don't we have to check for >= 0?